src/kernel/include/cap_ops.h

```diff
@@ -152,6 +163,184 @@ cap_cons(struct captbl *t, capid_t capto, capid_t capsub, capid_t expandid)
 	return ret;
 }
 
+/* PRY:this is dedicated to pgtbls */
+extern u32_t chal_MPU_update(struct pgtbl* table, u32_t Op_Flag);
+static inline int
+pgtbl_cons(struct captbl *t, capid_t capto, capid_t capsub, capid_t expandid)
+{
+	/*
+	 * Note: here we're relying on the fact that cap_captbl has an
+	 * identical layout to cap_pgtbl.
+	 */
+	struct cap_captbl *ct, *ctsub;
+	unsigned long *intern;
+	u32_t depth;
+	cap_t cap_type;
+	int ret = 0;
+
+	if (unlikely(capto == capsub)) return -EINVAL;
+	ct = (struct cap_captbl *)captbl_lkup(t, capto);
+	if (unlikely(!ct))             return -ENOENT;
+	cap_type = ct->h.type;
+	if (unlikely(cap_type != CAP_PGTBL)) return -EINVAL;
+	ctsub = (struct cap_captbl *)captbl_lkup(t, capsub);
+	if (unlikely(!ctsub))                    return -ENOENT;
+	if (unlikely(ctsub->h.type != cap_type)) return -EINVAL;
+
+	/* PRY:do not check this now */
+//	depth = ctsub->lvl;
+//	if (depth == 0) return -EINVAL; /* subtree must not have a root */
+	/* PRY: this assertion does not work for pgtbl. removed */
+//	assert(ct->captbl);
+//	if (cap_type == CAP_CAPTBL) {
+//		ret = captbl_cons(ct, ctsub, expandid);
+//	} else {
+		/* PRY: no need to care about cons/decons */
+		/* FIXME: we need to ensure TLB quiescence for pgtbl cons/decons! */
+		u32_t flags = 0, old_pte, new_pte, old_v, refcnt_flags;
+		u32_t old_flags;
+		u32_t size_order;
+
+		/* PRY:look up the position to add this to. If this is the top-level, we ignore than expandid and start direct insertion */
+		if(((((struct cap_pgtbl *)ct)->pgtbl.type_addr)&COS_PGTBL_TABLE)==0)
+		{
+			/* Do we have a lower level? */
+			intern=&(((struct cap_pgtbl *)ct)->pgtbl.type_addr);
+			if((((struct cap_pgtbl *)ct)->pgtbl.type_addr)!=0)
```

Comment:
This is due to the organization of the type_addr field. In the nodes that contain MPU data, the pointer points to the actual top-level node. This check is to make sure that there are no top-level nodes that are already cons'ed into it.